src/Compiler/pars.fsy

```diff
@@ -3649,6 +3649,47 @@ declExpr:
         let mWhileAll = unionRanges (rhs parseState 1) (rhs parseState 3)
         exprFromParseError (SynExpr.While (spWhile, arbExpr("whileGuard1", mWhileHeader), arbExpr("whileBody3", mWhileBodyArb), mWhileAll))  }
 
+  | WHILE_BANG declExpr doToken typedSequentialExprBlock doneDeclEnd 
+      { let mWhileHeader = unionRanges (rhs parseState 1) $2.Range
+        let spWhile = DebugPointAtWhile.Yes mWhileHeader 
+        let mWhileAll = unionRanges (rhs parseState 1) $4.Range
+        SynExpr.WhileBang (spWhile, $2, $4, mWhileAll) }
```

Comment:
We should probably extract these repeated rules into common ones for both `while` and `while!` expressions.

```
  | WHILE whileExprCore
    { ...  }

...

  | WHILE_BANG whileExprCore
    { ... }

...


whileExprCore:
  | declExpr doToken typedSequentialExprBlock doneDeclEnd
    { ... }

  | declExpr doToken typedSequentialExprBlock recover
    { ... }

  ...
```

Would be great to do it for other `!` variants of the expressions, like `do`, match`, and `let`.

Reply:
Good idea. The fact that there are 6 slightly different branches will make it annoying though.

Reply:
Please take a look. I think I've managed to minimize duplication here.

Reply:
@kerams It's perfect, thanks!